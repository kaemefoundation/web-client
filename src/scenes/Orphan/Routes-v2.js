import BasicInformation from "./components-v2/FormSections/BasicInformation";
import Family from "./components-v2/FormSections/Family";
import Medical from "./components-v2/FormSections/Medical";
import PostProfile from "./components-v2/FormSections/PostProfile";
import Followups from "./components-v2/FormSections/Followups";



const nav = [
  {
    path: "basic-information",
    component: BasicInformation,
    navName: "Basic Information"
  },
  {
    path: "family",
    component: Family,
    navName: "Family"
  },
  {
    path: "medical",
    component: Medical,
    navName: "Medical"
  },
  {
  	path:"post-profile",
  	component:PostProfile,
  	navName:"Post Profiling Information"
  },
  {
  	path:"followups",
  	component:Followups,
  	navName:"Follow Ups"
  }

  ];

  export default nav;
