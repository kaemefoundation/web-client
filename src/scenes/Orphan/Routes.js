import BasicInformation from "./components/FormSections/BasicInformation";
import IdentificationInformation from "./components/FormSections/IdentificationInformation";
import FamilyPart1 from "./components/FormSections/FamilyPart1";
import Residence from "./components/FormSections/Residence";
import AcademicHistory from "./components/FormSections/AcademicHistory";
import MedicalInformation from "./components/FormSections/MedicalInformation";
import SocialBehavior from "./components/FormSections/SocialBehavior";
import Emotional from "./components/FormSections/Emotional";
import Notes from "./components/FormSections/Notes";
import SocialBehaviorPart2 from "./components/FormSections/SocialBehaviorPart2";
import FamilyPart2 from "./components/FormSections/FamilyPart2";
import ResidencePart2 from "./components/FormSections/ResidencePart2";
import Notes2 from "./components/FormSections/Notes2";
import ChildDesignation from "./components/FormSections/ChildDesignation";
import ChildCareConference from "./components/FormSections/ChildCareConference";
import Followups from "./components/FormSections/Followups";


export const navPart1 = [
  {
    path: "basic-information",
    component: BasicInformation,
    navName: "Basic Information"
  },
  {
    path: "identification-information",
    component: IdentificationInformation,
    navName: "Identification Information"
  },
  { path: "family", component: FamilyPart1, navName: "Family" },
  { path: "residence", component: Residence, navName: "Residence History" },
  { path: "academic", component: AcademicHistory, navName: "Academic History" },
  {
    path: "medical-info",
    component: MedicalInformation,
    navName: "Medical Information"
  },
  {
    path: "social-behavior",
    component: SocialBehavior,
    navName: "Social/Behavioral Development"
  },
  { path: "emotional", component: Emotional, navName: "Emotional Functioning" },
  {
    path: "notes",
    component: Notes,
    navName: "Notes/Social Worker Observations"
  }
];
export const navPart2 = [
  {
    path: "social-behavior",
    component: SocialBehaviorPart2,
    navName: "Social/Behavioral Development"
  },
  { path: "family", component: FamilyPart2, navName: "Family" },
  {
    path: "residence",
    component: ResidencePart2,
    navName: "Residence (current & prior)"
  },
  {
    path: "notes",
    component: Notes2,
    navName: "Notes/Social Worker Observations"
  }
];
export const navPart3 = [
  {
    path: "child-designation",
    component: ChildDesignation,
    navName: "Child Designation"
  },
  {
    path: "child-care-conference",
    component: ChildCareConference,
    navName: "Child Care Conference"
  }
];
export const navPart4 = [
  { path: "followups", component: Followups, navName: "Follow Ups" }
];