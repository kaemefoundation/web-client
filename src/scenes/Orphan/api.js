export let devHttpEndpoint =
  "https://g2gp355fq6.execute-api.eu-central-1.amazonaws.com/development/";
export let localHttpEndpoint = "http://127.0.0.1:5000/orphans/";
export let currentEndpoint = devHttpEndpoint;


export function pick(o, ...fields) {
  return fields.reduce((a, x) => {
    if (o.hasOwnProperty(x)) a[x] = o[x];
    return a;
  }, {});
}
export function updateLocalStorage(key, data) {
  localStorage.setItem(key, JSON.stringify(data));
}
export function getLocalStorage(key) {
  return JSON.parse(localStorage.getItem(key));
}
export function getYears(startYear) {
  var currentYear = new Date().getFullYear(), years = [];
  startYear = startYear || 1990;
  while (startYear <= currentYear) {
    years.push({ value: startYear.toString(), label: startYear.toString() });
    startYear++;
  }
  return years;
}
export function newOrphan() {
  return {
    residences: [
      { is_current_residence: 1, reported_by: "orphanage" },
      { reported_by: "orphanage" },
      { reported_by: "orphan" },
      { reported_by: "orphan" }
    ],
    relationships: [
      { reported_by: "orphanage", relationship_type: "mother" },
      { reported_by: "orphanage", relationship_type: "father" },
      { reported_by: "orphan", relationship_type: "mother" },
      { reported_by: "orphan", relationship_type: "father" }
    ],
    vaccines: [
      { vaccine: "Tuberculosis (BCG)" },
      { vaccine: "Polio" },
      { vaccine: "Measles" },
      { vaccine: "Tetanus (TT)" },
      { vaccine: "Pentavalent" },
      { vaccine: "Yellow Fever (YF)" },
      { vaccine: "Rotavirus" },
      { vaccine: "Pneumonia (PCV-xx)" }
    ]
  };
}
export function getRegions() {
  let cachedRegions = getLocalStorage("regions");

  if (cachedRegions) {
    return Promise.resolve(cachedRegions);
  } else {
    return get(currentEndpoint + "regions").then(data => {
      let sortedData = data.sort();
      updateLocalStorage('regions',sortedData);
      return sortedData;
    });
  }
}
export function getOrphanages() {
  let cachedOrphanages = getLocalStorage("orphanages");
  if (cachedOrphanages) {
    return Promise.resolve(cachedOrphanages);
  } else {
    return get(currentEndpoint + "orphanages").then(data => {
      let sortedData = data.sort();
      updateLocalStorage('orphanages',sortedData);
      return sortedData;
    });
  }
}

export function getOrphanList() {
  if (window.navigator.onLine){ 
   
    let cachedOrphans = getLocalStorage("orphans");
   
    if(cachedOrphans){
      return Promise.resolve(cachedOrphans);
    }else{
      return get(currentEndpoint + "orphans/list").then((data)=>{
        updateLocalStorage("orphans",data);
        return data;
      });
    }
    
  } else {
    let orphanIdList = getLocalStorage("orphan-id-list") || [];
    let records = [];
    orphanIdList.map((id)=>{
      records.push(getLocalStorage(id));
      return id;
    });
    return Promise.resolve({ records: records });
  }
}
export function addToOrphanIdList(orphanId){
  let localOrphanIdList = getLocalStorage("orphan-id-list") || [];
  localOrphanIdList.push(orphanId);
  updateLocalStorage("orphan-id-list",localOrphanIdList);
}
export function getOrphanData(childId, relatedTable) {
  if (window.navigator.onLine) {
    return get(currentEndpoint + "orphans/" + childId);
  } else {
    let orphan = getLocalStorage(childId) || {};
    return Promise.resolve(orphan);
  }
}
export function putOrphanData(data) {

  updateLocalStorage(data.id,data);

  if(window.navigator.onLine){
    return post(currentEndpoint + "orphans/" + data.id, data);
  }else{
    return Promise.resolve(data);
  }
  
}
export function post(url, data) {
  return fetch(url, { method: "post", body: JSON.stringify(data) })
    .then(response => {
      return handleResponse(response, url);
    })
    .catch(error => {
      return Promise.reject(new Error(error.message));
    });
}
export function get(url) {
  return fetch(url)
    .then(response => {
      return handleResponse(response, url);
    })
    .catch(error => {
      return Promise.reject(new Error(error.message));
    });
}

export function handleResponse(response, url) {
  if (response.ok) {
    const contentType = response.headers.get("Content-Type") || "";

    if (contentType.includes("application/json")) {
      return response.json().catch(error => {
        return Promise.reject(new Error("Invalid JSON: " + error.message));
      });
    }

    if (contentType.includes("text/html")) {
      return response
        .text()
        .then(html => {
          return {
            page_type: "generic",
            html: html
          };
        })
        .catch(error => {
          return Promise.reject(new Error("HTML error: " + error.message));
        });
    }

    return Promise.reject(new Error("Invalid content type: " + contentType));
  }

  if (response.status === 404) {
    return Promise.reject(new Error("Page not found: " + url));
  }
  if (response.status === 500) {
    return Promise.reject(new Error("500 error: " + url));
  }
  return Promise.reject(new Error("HTTP error: " + response.status));
}
