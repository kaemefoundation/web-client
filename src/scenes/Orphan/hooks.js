import React, { useState, useEffect} from "react";


import{
 getRegions,
  getOrphanages,
  getData
} from "./api.js";


export const OrphanContext = React.createContext({onSave:function(){},child:{},formClass:"",onClickLoadData:function(){}});

export function useOrphanages(){
  let [orphanages,updateOrphanages] = useState([]);

  useEffect(()=>{
     getOrphanages().then((data) =>
     { updateOrphanages(data);});
    },[orphanages.length]);
  
  return [orphanages,updateOrphanages];
}
export function useRegions(){
  let [regions, updateRegions] = useState([]);
  useEffect(()=>{
     getRegions().then((data) =>
     { updateRegions(data);});
    },[regions.length]);
  
  return [regions,updateRegions];
}
export function useRegionsAndOrphanages() {
  let [regions, updateRegions] = useState([]);
  let [orphanages, updateOrphanages] = useState([]);
  useEffect(() => {
    let promises = getData([getRegions(), getOrphanages()]);
    promises.then(values => {
      updateRegions(values[0].value);
      updateOrphanages(values[1].value);
    });
  }, [regions.length, orphanages.length]);
  return [regions, orphanages];
}

