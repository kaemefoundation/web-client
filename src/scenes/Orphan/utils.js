export function dropdownListSort(a,b){
	return a.label.localeCompare(b.label);
}

export function updateLocalStorage(key, data) {
  localStorage.setItem(key, JSON.stringify(data));
}
export function getLocalStorage(key) {
  return JSON.parse(localStorage.getItem(key));
}

export function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
export function onlineOffline(callbackForOnline,callbackForOffline){
	if(window.navigator.onLine){
		console.log("online");
		return callbackForOnline();
	}else{
		console.log("offline");
		return callbackForOffline();
	}
}
export function getCurrentOrphanageIndex(data){
	let index = data.findIndex((element)=>{
		return element.is_current_residence === 1;
	});
	return index;
}

export function getPriorResidenceIndex(data,reportedBy){
	let index = data.findIndex((element)=>{
		return element.reported_by === reportedBy && element.is_current_residence !== 1;
	});
	return index;
}
export function getOtherResidences(data,priorResidenceIndex,reportedBy){
	
	let indexes = [];
	
	data.map((element,index)=>{
		if(element.is_current_residence !== 1 && index !== priorResidenceIndex && element.reported_by === reportedBy){
			indexes.push(index);
		}
		return element;
	});
	return indexes;
}
export function reportedByOrphanFather(element) {
	return findRelationshipIndex("father", "orphan", element);
}
export function reportedByOrphanMother(element) {
	return findRelationshipIndex("mother", "orphan", element);
}
export function reportedByOrphanageFather(element) {
	return findRelationshipIndex("father", "orphanage", element);
}
export function reportedByOrphanageMother(element) {
	return findRelationshipIndex("mother", "orphanage", element);
}
export function findMultipleRelationshipIndexes(
	data,
	reportedBy,
	includeTypes,
	excludeTypes
) {
	let indexes = [];

	data.map((element, index) => {
		if (element.reported_by === reportedBy) {
			if (includeTypes.length > 0) {
				if (
					includeTypes.includes(element.relationship_type) &&
					!excludeTypes.includes(element.relationship_type)
				) {
					indexes.push(index);
				}
			} else if (!excludeTypes.includes(element.relationship_type)) {
				indexes.push(index);
			}
		}
		return element;
	});
	return indexes;
}
export function findRelationshipIndex(type, reportedBy, element) {
	return (
		element.reported_by === reportedBy && element.relationship_type === type
	);
}
export const  units = {'zero': 0,
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9,
    'ten': 10,
    'eleven': 11,
    'twelve': 12,
    'thirteen': 13,
    'fourteen': 14,
    'fifteen': 15,
    'sixteen': 16
}
