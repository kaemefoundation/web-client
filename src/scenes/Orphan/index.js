//3rd party library components
import React, { Component } from "react";
import { Route, Switch } from "react-router-dom";

//Kaeme components/functions
import Navigation from "./components/Navigation";
import OrphanList from "./components/OrphanList";
import { navPart1, navPart2, navPart3, navPart4 } from "./Routes";
import { getOrphanData, putOrphanData,newOrphan,updateLocalStorage,addToOrphanIdList } from "./api.js";

class Orphan extends Component {
  constructor(props) {
    super(props);
    this.state = { child: {}, error: false,loading:false };
    this.saveChild = this.saveChild.bind(this);
    this.updateChildWithRelatedData = this.updateChildWithRelatedData.bind(this);
    this.addChild = this.addChild.bind(this);
  }
  
  saveChild(updatedChild) {
    
    let newChild = Object.assign({}, updatedChild, {
      id: this.state.child.id
    });
    this.setState({loading:true},()=>{
      putOrphanData(newChild)
      .then((data)=>{
        this.setState({child:data,loading:false});
      })
      .catch(() => {
        this.setState({ error: true });
      });
    });
    
  }
  addChild(){
    localStorage.removeItem("orphans");
    putOrphanData(newOrphan()).then((data)=>{
      this.setState({child:data},()=>{
        addToOrphanIdList(data);
        window.location.href='/orphan/'+data.id+'/part1/basic-information';
      })
    });
  }
  updateChildWithRelatedData(relatedTableName,data){
    let relatedData = {};
    relatedData[relatedTableName] = data;
    this.setState({child: Object.assign({},this.state.child,relatedData)});
    
  }
  componentDidMount() {
    let childId = this.props.match.params.id;
    if (childId) {
      this.setState({loading:true},()=>{
        getOrphanData(childId)
        .then(data => {
          this.setState({ child: data,loading:false });
        })
        .catch(() => {
          this.setState({ error: true,loading:false });
        });
      });
      
    }
  }
  render() {
    let formClass = "ui attached form";
    formClass = this.state.loading ? "ui form loading" : formClass;
    return (
      <div className="ui grid">
        {this.props.match.params.id
          ? <Navigation id={this.props.match.params.id} />
          : ""}
          
        <Switch>
          <Route path="/" exact render={props=>(<OrphanList addChild={this.addChild} />)} />
          <Route path="/orphan" exact render={props=>(<OrphanList addChild={this.addChild} />)} />
         
          {navPart1.map((route, index) => (
              <Route
                key={"part1" + index}
                path={"/orphan/:id/part1/" + route.path}
                render={props => (
                  <route.component
                    child={this.state.child}
                    onSave={this.saveChild}
                    updateChildWithRelatedData={this.updateChildWithRelatedData}
                    formClass={formClass}
                  />
                )}
              />
            ))}
          {navPart2.map((route, index) => (
              <Route
                key={"part2" + index}
                path={"/orphan/:id/part2/" + route.path}
                render={props => (
                  <route.component
                    child={this.state.child}
                    onSave={this.saveChild}
                    updateChildWithRelatedData={this.updateChildWithRelatedData}
                    formClass={formClass}
                  />
                )}
              />
            ))}
          {navPart3.map((route, index) => (
              <Route
                key={"part3"+index}
                path={"/orphan/:id/part3/" + route.path}
                render={props => (
                  <route.component
                    child={this.state.child}
                    onSave={this.saveChild}
                    updateChildWithRelatedData={this.updateChildWithRelatedData}
                    formClass={formClass}
                  />
                )}
              />
            ))}
          {navPart4.map((route, index) => (
              <Route
                key={"part4" + index}
                path={"/orphan/:id/part4/" + route.path}
                render={props => (
                  <route.component
                    child={this.state.child}
                    onSave={this.saveChild}
                    updateChildWithRelatedData={this.updateChildWithRelatedData}
                    formClass={formClass}
                  />
                )}
              />
            ))}
          
        </Switch>
      </div>
    );
  }
}

export default Orphan;
