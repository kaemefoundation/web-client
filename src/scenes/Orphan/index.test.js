import React from 'react';
import ReactDOM from 'react-dom';
import Orphan from './index.js';
import renderer from 'react-test-renderer';


test('Orphan component matches snapshot', () => {
    const orphan = renderer.create(<Orphan/>);
    let tree = orphan.toJSON();
    expect(tree).toMatchSnapshot();
  });