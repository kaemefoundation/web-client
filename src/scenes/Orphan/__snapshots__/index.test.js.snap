exports[`test Orphan component matches snapshot 1`] = `
<div>
  Orphan is changed
</div>
`;
